File: core/src/com/alta_v2/physics/utils/TiledMapParser.java
package com.alta_v2.physics.utils;

import com.alta_v2.physics.executionContext.altitude.AltitudeMap;
import com.alta_v2.physics.executionContext.altitude.AltitudeMapPoint;
import com.alta_v2.physics.executionContext.altitude.PointAvailability;
import com.badlogic.gdx.assets.AssetManager;
import com.badlogic.gdx.maps.tiled.TiledMap;
import com.badlogic.gdx.maps.tiled.TiledMapTileLayer;
import com.badlogic.gdx.maps.tiled.TmxMapLoader;
import lombok.experimental.UtilityClass;

/**
 * Provides the parser for tiled map.
 */
@UtilityClass
public class TiledMapParser {

    public AltitudeMap parse(String path) {
        TiledMap tiledMap = loadMap(path);

        int countX = tiledMap.getProperties().get("width", Integer.class);
        int countY = tiledMap.getProperties().get("height", Integer.class);

        try {
            return AltitudeMap.builder()
                    .horizontalTilesCount(countX)
                    .verticalTilesCount(countY)
                    .tileWidth(tiledMap.getProperties().get("tilewidth", Integer.class))
                    .tileHeight(tiledMap.getProperties().get("tileheight", Integer.class))
                    .altitudes(getAltitudes(tiledMap, countX, countY))
                    .build();
        } finally {
            tiledMap.dispose();
        }
    }

    private TiledMap loadMap(String path) {
        AssetManager assetManager = new AssetManager();
        assetManager.setLoader(TiledMap.class, new TmxMapLoader());
        assetManager.load(path, TiledMap.class);
        assetManager.finishLoading();

        TiledMap tiledMap = assetManager.get(path, TiledMap.class);
        assetManager.dispose();
        return tiledMap;
    }

    private AltitudeMapPoint[][] getAltitudes(TiledMap tiledMap, int countX, int countY) {
        TiledMapTileLayer barrierLayer = (TiledMapTileLayer) tiledMap.getLayers().get("over_barrier");
        AltitudeMapPoint[][] altitudes = new AltitudeMapPoint[countX][countY];

        for (int i = 0; i < countX; i++) {
            for (int j = 0; j < countY; j++) {
                if (barrierLayer.getCell(i, j) == null) {
                    altitudes[i][j] = new AltitudeMapPoint(PointAvailability.FREE);
                } else {
                    altitudes[i][j] = new AltitudeMapPoint(PointAvailability.BARRIER);
                }
            }
        }

        return altitudes;
    }
}

File: core/src/com/alta_v2/physics/utils/TiledMapParser.kt
package com.alta_v2.physics.utils

import com.alta_v2.physics.executionContext.altitude.AltitudeMap
import com.alta_v2.physics.executionContext.altitude.AltitudeMapPoint
import com.alta_v2.physics.executionContext.altitude.PointAvailability
import com.alta_v2.utils.MAP_HEIGHT
import com.alta_v2.utils.MAP_WIDTH
import com.alta_v2.utils.TILE_HEIGHT
import com.alta_v2.utils.TILE_WIDTH
import com.badlogic.gdx.assets.AssetManager
import com.badlogic.gdx.maps.tiled.TiledMap
import com.badlogic.gdx.maps.tiled.TiledMapTileLayer
import com.badlogic.gdx.maps.tiled.TmxMapLoader

/**
 * Provides the parser for tiled map.
 */
class TiledMapParser {

    companion object {
        fun parse(path: String): AltitudeMap {
            val tiledMap = loadMap(path)
            val countX = tiledMap.properties.get(MAP_WIDTH, Int::class.java)
            val countY = tiledMap.properties.get(MAP_HEIGHT, Int::class.java)
            return try {
                AltitudeMap(
                        tileWidth = tiledMap.properties.get(TILE_WIDTH, Int::class.java),
                        tileHeight = tiledMap.properties.get(TILE_HEIGHT, Int::class.java),
                        altitudes = getAltitudes(tiledMap, countX, countY)
                )
            } finally {
                tiledMap.dispose()
            }
        }

        private fun loadMap(path: String): TiledMap {
            val assetManager = AssetManager()
            assetManager.setLoader(TiledMap::class.java, TmxMapLoader())
            assetManager.load(path, TiledMap::class.java)
            assetManager.finishLoading()
            val tiledMap = assetManager.get(path, TiledMap::class.java)
            assetManager.dispose()
            return tiledMap
        }

        private fun getAltitudes(tiledMap: TiledMap, countX: Int, countY: Int): Array<Array<AltitudeMapPoint>> {
            val barrierLayer = tiledMap.layers["over_barrier"] as TiledMapTileLayer
            // val altitudes = Array(countX) { arrayOfNulls<AltitudeMapPoint>(countY) }

            return Array(countX) {i ->
                Array(countY) { j ->
                    if (barrierLayer.getCell(i, j) == null) {
                        AltitudeMapPoint(PointAvailability.FREE)
                    } else {
                        AltitudeMapPoint(PointAvailability.BARRIER)
                    }
                }
            }
        }
    }
}